crates/libs/registry/src/key_iterator.rs

```diff
@@ -3,66 +3,59 @@ use super::*;
 /// An iterator of registry key names.
 pub struct KeyIterator<'a> {
     key: &'a Key,
-    range: core::ops::Range<usize>,
+    idx: u32,
     name: Vec<u16>,
 }
 
 impl<'a> KeyIterator<'a> {
     pub(crate) fn new(key: &'a Key) -> Result<Self> {
-        let mut count = 0;
-        let mut max_len = 0;
+        let info = get_key_info(key)?;
 
-        let result = unsafe {
-            RegQueryInfoKeyW(
-                key.0,
-                null_mut(),
-                null_mut(),
-                null_mut(),
-                &mut count,
-                &mut max_len,
-                null_mut(),
-                null_mut(),
-                null_mut(),
-                null_mut(),
-                null_mut(),
-                null_mut(),
-            )
-        };
-
-        win32_error(result).map(|_| Self {
+        Ok(Self {
             key,
-            range: 0..count as usize,
-            name: vec![0; max_len as usize + 1],
+            idx: 0,
+            name: vec![0; (info.subkey_name_max + 1) as usize],
         })
     }
+
+    fn resize(&mut self) -> Result<()> {
+        let info = get_key_info(self.key)?;
+        self.name.resize((info.subkey_name_max + 1) as usize, 0);
+        Ok(())
+    }
 }
 
 impl<'a> Iterator for KeyIterator<'a> {
     type Item = String;
 
     fn next(&mut self) -> Option<Self::Item> {
-        self.range.next().and_then(|index| {
-            let mut len = self.name.len() as u32;
-
-            let result = unsafe {
-                RegEnumKeyExW(
-                    self.key.0,
-                    index as u32,
-                    self.name.as_mut_ptr(),
-                    &mut len,
-                    null(),
-                    null_mut(),
-                    null_mut(),
-                    null_mut(),
-                )
-            };
+        let mut len = self.name.len() as u32;
+        let result = unsafe {
+            RegEnumKeyExW(
+                self.key.0,
+                self.idx,
+                self.name.as_mut_ptr(),
+                &mut len,
+                null_mut(),
+                null_mut(),
+                null_mut(),
+                null_mut(),
+            )
+        };
 
-            if result != 0 {
-                debug_assert_eq!(result, ERROR_NO_MORE_ITEMS);
-                None
-            } else {
-                Some(String::from_utf16_lossy(&self.name[0..len as usize]))
+        if result == ERROR_MORE_DATA {
+            if self.resize().is_err() {
+                return None;
             }
-        })
+            return self.next();
```

Comment:
This is recursive - not necessarily bad in this scenario but worth calling out if it's not obvious.